udf-compiler/src/main/scala/com/nvidia/spark/udf/Instruction.scala

```diff
@@ -73,6 +73,39 @@ private object Repr {
     var string: Expression = Literal.default(StringType)
   }
 
+  case class DateTimeFormatter private (private[Repr] val pattern: Expression) extends CompilerInternal("java.time.format.DateTimeFormatter") {
+    def invoke(methodName: String, args: List[Expression]): Expression = {
+      methodName match {
+        case _ =>
+          throw new SparkException(s"Unsupported DateTimeFormatter op ${methodName}")
+      }
+    }
+  }
+  object DateTimeFormatter {
+    private def apply(pattern: Expression): DateTimeFormatter = new DateTimeFormatter(pattern)
+    def ofPattern(pattern: Expression): DateTimeFormatter = DateTimeFormatter(pattern)
+  }
+
+  case class LocalDateTime private (private val dateTime: Expression) extends CompilerInternal("java.time.LocalDateTime") {
```

Comment:
As far as I understand, a `java.time.LocalDateTime` doesn't have a timezone. 

The GPU portion of the plugin only enables date/time acceleration if Java's timezone is UTC. The assumption is the user knows what they are doing, and so they need to ensure that the Java timezone is set as such.

Should we put this under a similar protection? Since when replacing `getHour` with `Hour` we've made code that used to be timezone agnostic, all of the sudden care about the java timezone. Please correct me if you've already tested this and is not an issue.

Note here's an example for `Hour` in the GPU acceleration case:

https://github.com/NVIDIA/spark-rapids/blob/ce1f9b8b6201057a37a1a4add87f1ba0ec15c5bb/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala#L916

Reply:
Please correct me if I am wrong.
But `None` is passed as `timeZoneId` to Hour as we are not setting any specific time zone ID.
Wouldn't it keep the expression timezone agnostic?

Reply:
Looks like `ResolveTimeZone` is setting `timeZoneId` in that scenario. This comes from `spark.sql.session.timeZone`. which is `TimeZone.getDefault.getID`. 

So I think this is UTC as long as the Java side doesn't have that timezone. If you have a different timezone there, then it would take that timezone.

I am not sure that any of the functions implemented here would change given a different timezone, so this may be a non issue @seanprime7

Reply:
Okay.  I have just added a test case with timezoned datetime strings.
It indeed shows that `Hour` doesn't ignore the timezone and performs a timezone conversion.
I need to add a check to prevent it.

Reply:
I have added the check against timezoned datetime format.